homework_2/offerfeed/internal/infrastructure/http/feedMerger.go

```diff
@@ -0,0 +1,45 @@
+package http
+
+import (
+	"context"
+	"log"
+
+	"code-cadets-2021/homework_2/offerfeed/internal/domain/models"
+)
+
+type FeedMerger struct {
+	Feeds   []Feed
+	Updates chan models.Odd
+}
+
+func NewFeedMerger(feeds ...Feed) *FeedMerger {
+	return &FeedMerger{Feeds: feeds, Updates: make(chan models.Odd)}
+}
+
+func (a *FeedMerger) Start(ctx context.Context) error {
+	defer close(a.Updates)
+	defer log.Printf("shutting down %s", a)
+
+	for _, feed := range a.Feeds {
+		feed.SetUpdates(a.Updates)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
```

Comment:
Usually you want to perform what's called a fan-in when you have multiple channels. Usually you never push a single channel to be filled by multiple feeds, you want to integrate those feeds into a single feed which you then offer for consumption. You do this is by reading each feed channel and putting their output into a single channel which you made in this struct. 
So, why would you do this instead of the other way around?
Think about how would your feeds notify you that they are closed? They couldn't, they would only be closed when the context is done and this channel here is closed. You might wait for a long time for that to happen, without reason. You want for each feed to be independent and used on its own. If you didn't want to merge feeds, think where you would have to create the channel which you would inject into the feed. 
So what I propose is for each feed to have its own channel, which you read here in separate threads (using `go func`) or by using a select statement (https://gobyexample.com/select) and put the result into a channel of the `feed merger`. 
If you are interested in more, you can find some other concurrency patterns here https://blog.golang.org/pipelines

Reply:
If I were to use the _fan-in_ pipeline, how would I also wait for the _context_ to end so the channel can be closed?

Reply:
Also, should I start separate feeds in _feedMerger_, or they should be stared as a _task_, but before the _feedMerger_?

Reply:
Also, if I were to use the _select_ statement, is it possible to have an arbitrary number of _feeds_ or the number should be fixed, as there is a fixed number of _cases_ in the _select_ statement.

Reply:
> If I were to use the _fan-in_ pipeline, how would I also wait for the _context_ to end so the channel can be closed?

I think I understand this one. When the `context` closes, separate feeds will close and so will the `feedMerger`.

Reply:
You answered your first question already I see. You are correct, since they all share the same context, they will close in succession. 

>Also, should I start separate feeds in feedMerger, or they should be stared as a task, but before the feedMerger?
Feeds should be started as tasks like you already did in your `main` function 👍 

> Also, if I were to use the select statement, is it possible to have an arbitrary number of feeds or the number should be fixed, as there is a fixed number of cases in the select statement?
Yes, unfortunately, as far as I know you cannot have an arbitrary number of feeds with a select statement. To consume an arbitrary number of feeds you would need to spawn a goroutine for each feed that reads it and outputs the result into the shared feed merger channel. 
To go way further, technically there is a way to do it, but not with the vanilla select, but using select from the `reflect` package. But this deals with runtime reflection (which is pretty advanced stuff) and usually you want to avoid it if it is not necessary since you can cause a `panic` quite easily. So, in the context of this course, the answer is no 😁

Reply:
Alright, I pushed all the updates, except for the name change in NotAJsonFeed. I will do that later today and maybe play a little with different timeouts, as you mentioned above. Everything else should now be fixed.